### PR TITLE
Refactor `TrustedHostsNetworkResolver`

### LINKED ISSUE #15: Fix remaining mutants

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | See "Development" section. Additionally fixes #15, #21.

### CODE REVIEW

File: tests/TrustedHostsNetworkResolverTest.php [hunk 1]
```diff
@@ -2,847 +2,2580 @@
...
-        $this->assertSame($expectedPath, $requestHandler->processedRequest->getUri()->getPath());
-        $this->assertSame($expectedQuery, $requestHandler->processedRequest->getUri()->getQuery());
-        $this->assertSame($expectedPort, $requestHandler->processedRequest->getUri()->getPort());
+        $this->assertNotSame($middleware, $middleware->withConnectionChainItemsAttribute('connectionChainItems'));
     }
 
-    public function dataProcessNotTrusted(): array
+    public function dataWithTypicalForwardedHeaders(): array
     {
         return [
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:423]: Just in case we'll have tests performance issue in the future, we can use `yield` in similar cases which is going to prevent instantiating data for all cases all at once.

File: README.md [hunk 2]
```diff
@@ -61,58 +61,270 @@ $response = $middleware->process($request, $handler);
 
 ### `TrustedHostsNetworkResolver`
 
-Trusted hosts network resolver can set IP, protocol, host, URL, and port based on trusted headers such as
-`Forward` or `X-Forwarded-Host` coming from trusted hosts you define. Usually these are load balancers.
+Scans the entire connection chain and resolves the data from forwarded headers taking into account trusted IPs.
```
[the remark this answers is not in the document]
Comment (arogachev) [reply README.md:64]: https://datatracker.ietf.org/doc/html/rfc7239#section-7.5

File: tests/TrustedHostsNetworkResolverTest.php (hunk 1 above)
Comment (arogachev) [reply tests/TrustedHostsNetworkResolverTest.php:423]: Good idea! Applied it in data providers where objects are created.

File: README.md [hunk 3]
```diff
@@ -61,58 +61,270 @@ $response = $middleware->process($request, $handler);
...
+[obfuscated](https://datatracker.ietf.org/doc/html/rfc7239#section-6.2). Used with `Forwarded` RFC header. 
+
+The typical use case is load balancers.
+
+#### Trusted IPs
+
+A list of trusted IPs from connection chain.
+
+This is the only requirement, the rest of the settings are optional. Trusted IPs must contain the value in
+`$_SERVER['REMOTE_ADDR']`. For example, for `18.18.18.18` server's remote address and 2 trusted proxies - `2.2.2.2` and
```
[the remark this answers is not in the document]
Comment (arogachev) [reply README.md:82]: https://datatracker.ietf.org/doc/html/rfc7239#section-7.5

File: tests/TrustedHostsNetworkResolverTest.php [hunk 4]
```diff
@@ -2,847 +2,2580 @@
...
+                    serverParams: ['REMOTE_ADDR' => '18.18.18.18'],
+                ),
+                [
+                    'requestClientIp' => '5.5.5.5',
+                    'removedHeaders' => [
+                        'X-Forwarded-For',
+                        'X-Forwarded-Proto',
+                        'X-Forwarded-Host',
+                        'X-Forwarded-Port',
+                        'Front-End-Https',
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:464]: Should we remove `Forwarded` as well since it's not in the group that matched?

File: tests/TrustedHostsNetworkResolverTest.php [hunk 5]
```diff
@@ -2,847 +2,2580 @@
...
+                        'Y-Forwarded-Proto' => ['https'],
+                        'Y-Forwarded-Host' => ['example3.com'],
+                        'Y-Forwarded-Port' => ['8030'],
+                        'Front-End-Https' => 'On',
+                        'Non-Forwarded' => 'test',
+                    ],
+                    serverParams: ['REMOTE_ADDR' => '18.18.18.18'],
+                ),
+                [
+                    'requestClientIp' => '18.18.18.18',
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:518]: Shouldn't it be 1.1.1.1?

IPs are from `X-FORWARDED-FOR` and `REMOTE_ADDR`:

```
                                    trusted
'9.9.9.9', '5.5.5.5', '1.1.1.1', '18.18.18.18'
                          ↑ ← ← ← ← ← ↓                     
```

File: tests/TrustedHostsNetworkResolverTest.php [hunk 6]
```diff
@@ -2,847 +2,2580 @@
...
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:578]: Shouldn't it be 1.1.1.1:

```
                                    trusted
'9.9.9.9', '5.5.5.5', '1.1.1.1', '18.18.18.18'
                          ↑ ← ← ← ← ← ↓                     
```

File: tests/TrustedHostsNetworkResolverTest.php [hunk 7]
```diff
@@ -2,847 +2,2580 @@
...
+                ],
+            ],
+            yield 'RFC header, empty' => [
+                $this
+                    ->createMiddleware()
+                    ->withTrustedIps(['8.8.8.8', '19.19.19.19'])
+                    ->withConnectionChainItemsAttribute('connectionChainItems'),
+                $this->createRequest(
+                    headers: ['Forwarded' => ''],
+                    serverParams: ['REMOTE_ADDR' => '18.18.18.18'],
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:753]: Both are equivalent since REMOTE_ADDR is not trusted. Same for similar cases above.

File: tests/TrustedHostsNetworkResolverTest.php [hunk 8]
```diff
@@ -2,847 +2,2580 @@
...
+                    headers: ['Forwarded' => ['for=9.9.9.9', 'for=5.5.5.5', 'for=2.2.2.2']],
+                    serverParams: ['REMOTE_ADDR' => '18.18.18.18'],
+                ),
                 [
-                    'ipHeaders' => [
-                        [TrustedHostsNetworkResolver::IP_HEADER_TYPE_RFC7239, 'header1'],
-                        'x-forwarded-for',
-                        1,
-                        [TrustedHostsNetworkResolver::IP_HEADER_TYPE_RFC7239, 'header2'],
+                    'requestClientIp' => '18.18.18.18',
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:799]: Shouln't it be 2.2.2.2?

File: tests/TrustedHostsNetworkResolverTest.php [hunk 9]
```diff
@@ -2,847 +2,2580 @@
...
+                $this->createRequest(
+                    headers: ['X-Forwarded-For' => ['9.9.9.9', '5.5.5.5', '2.2.2.2']],
+                    serverParams: ['REMOTE_ADDR' => '18.18.18.18'],
+                ),
                 [
-                    'ipHeaders' => [
-                        [TrustedHostsNetworkResolver::IP_HEADER_TYPE_RFC7239, 'header1'],
-                        [TrustedHostsNetworkResolver::IP_HEADER_TYPE_RFC7239],
-                        [TrustedHostsNetworkResolver::IP_HEADER_TYPE_RFC7239, 'header2'],
+                    'requestClientIp' => '18.18.18.18',
```
Comment (samdark) [inline tests/TrustedHostsNetworkResolverTest.php:824]: Shouldn't it be 2.2.2.2?

File: tests/TrustedHostsNetworkResolverTest.php (hunk 4 above)
Comment (arogachev) [reply tests/TrustedHostsNetworkResolverTest.php:464]: https://github.com/yiisoft/proxy-middleware/blob/bcacf039788a469a35e84fb8ca4e153d13adb1fe/tests/TrustedHostsNetworkResolverTest.php#L435

https://github.com/yiisoft/proxy-middleware/blob/bcacf039788a469a35e84fb8ca4e153d13adb1fe/README.md?plain=1#L237

> in forwarded header groups

Do you mean it should be only 1 matching forwarded header group instead?

File: tests/TrustedHostsNetworkResolverTest.php (hunk 6 above)
Comment (arogachev) [reply tests/TrustedHostsNetworkResolverTest.php:578]: The same as https://github.com/yiisoft/proxy-middleware/pull/20#discussion_r1236772687.

File: tests/TrustedHostsNetworkResolverTest.php (hunk 8 above)
Comment (arogachev) [reply tests/TrustedHostsNetworkResolverTest.php:799]: The same as in https://github.com/yiisoft/proxy-middleware/pull/20#discussion_r1236772687.

File: tests/TrustedHostsNetworkResolverTest.php (hunk 9 above)
Comment (arogachev) [reply tests/TrustedHostsNetworkResolverTest.php:824]: The same as in https://github.com/yiisoft/proxy-middleware/pull/20#discussion_r1236772687.

File: tests/TrustedHostsNetworkResolverTest.php (hunk 7 above)
Comment (arogachev) [reply tests/TrustedHostsNetworkResolverTest.php:753]: > Both are equivalent

What do you mean by "both"?

> Same for similar cases above.

What cases exactly?

> REMOTE_ADDR is not trusted

It tests code that is executed before dealing with trusted IPs. 

For this case in particular it's this check: 

https://github.com/yiisoft/proxy-middleware/blob/bcacf039788a469a35e84fb8ca4e153d13adb1fe/src/TrustedHostsNetworkResolver.php#L420

Removing this test case causes appearing of escaped mutant.

Screenshot from 2023-06-21 16-41-32

File: tests/TrustedHostsNetworkResolverTest.php (hunk 4 above)
Comment (samdark) [reply tests/TrustedHostsNetworkResolverTest.php:464]: Yes. Leave the headers from the group that matched, remove the rest.

File: tests/TrustedHostsNetworkResolverTest.php (hunk 7 above)
Comment (samdark) [reply tests/TrustedHostsNetworkResolverTest.php:753]: > What do you mean by "both"?

Both "RFC header, empty" and previous case.

Comment (samdark) [reply tests/TrustedHostsNetworkResolverTest.php:753]: > Removing this test case causes appearing of escaped mutant.

Yes but the case is artificial. It would make sense only if the proxy is trusted.

File: README.md [hunk 10]
```diff
@@ -2,18 +2,18 @@
...
+[![Build status](https://github.com/yiisoft/proxy-middleware/workflows/build/badge.svg)](https://github.com/yiisoft/proxy-middleware/actions?query=workflow%3Abuild)
+[![Code Coverage](https://codecov.io/gh/yiisoft/proxy-middleware/branch/master/graph/badge.svg)](https://codecov.io/gh/yiisoft/proxy-middleware)
+[![Mutation testing badge](https://img.shields.io/endpoint?style=flat&url=https%3A%2F%2Fbadge-api.stryker-mutator.io%2Fgithub.com%2Fyiisoft%2Fproxy-middleware%2Fmaster)](https://dashboard.stryker-mutator.io/reports/github.com/yiisoft/proxy-middleware/master)
+[![static analysis](https://github.com/yiisoft/proxy-middleware/workflows/static%20analysis/badge.svg)](https://github.com/yiisoft/proxy-middleware/actions?query=workflow%3A%22static+analysis%22)
+[![type-coverage](https://shepherd.dev/github/yiisoft/proxy-middleware/coverage.svg)](https://shepherd.dev/github/yiisoft/proxy-middleware)
+[![psalm-level](https://shepherd.dev/github/yiisoft/proxy-middleware/level.svg)](https://shepherd.dev/github/yiisoft/proxy-middleware)
 
 The package ...
```
Comment (samdark) [inline README.md:18]: Missing part. But we can do it in a separate PR not to hold the merge.

File: src/TrustedHostsNetworkResolver.php [hunk 11]
```diff
@@ -13,732 +12,776 @@
...
+
+            $this->assertIsAllowedProtocol(
+                $protocol,
+                'Value returned from callable for protocol header',
+                inRuntime: true,
+            );
         }
 
-        return [null, null, []];
+        // TODO: Add validation flag.
```
Comment (samdark) [inline src/TrustedHostsNetworkResolver.php:499]: ?

**Review by samdark — APPROVED** [review]

File: README.md (hunk 10 above)
Comment (arogachev) [reply README.md:18]: #23

File: src/TrustedHostsNetworkResolver.php (hunk 11 above)
Comment (arogachev) [reply src/TrustedHostsNetworkResolver.php:499]: I removed it because it's not actual anymore.
